File: src/211.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// https://leetcode.cn/problems/design-add-and-search-words-data-structure/solutions/1440451/c-by-codee-3-nka1/
#define MAX_CHAR (26)  // word in addWord consists of lowercase English letters.
typedef struct WordDictionary {
    struct WordDictionary* child[MAX_CHAR];
    bool isEnd;
} WordDictionary;
WordDictionary* wordDictionaryCreate() {
    WordDictionary* pRetVal = (WordDictionary*)malloc(sizeof(WordDictionary));
    if (pRetVal == NULL) {
        perror("malloc");
        return pRetVal;
    }
    memset(pRetVal->child, 0, MAX_CHAR * sizeof(WordDictionary*));
    pRetVal->isEnd = false;

    return pRetVal;
}
void wordDictionaryAddWord(WordDictionary* obj, char* word) {
    int len = strlen(word);

    int ch;
    int i;
    for (i = 0; i < len; ++i) {
        ch = word[i] - 'a';
        if (obj->child[ch] != NULL) {
            obj = obj->child[ch];
            continue;
        }
        obj->child[ch] = (WordDictionary*)malloc(sizeof(WordDictionary));
        obj = obj->child[ch];
        memset(obj->child, 0, sizeof(obj->child));
        obj->isEnd = false;
    }

    obj->isEnd = true;
}
bool wordDictionarySearch(WordDictionary* obj, char* word) {
    bool retVal = false;

    if (obj == NULL) {
        return retVal;
    }

    int len = strlen(word);
    if (len == 0) {
        retVal = obj->isEnd;
        return retVal;
    }

    if (word[0] != '.') {
        retVal = wordDictionarySearch(obj->child[word[0] - 'a'], word + 1);
        return retVal;
    }

    int i;
    for (i = 0; i < MAX_CHAR; i++) {
        if (wordDictionarySearch(obj->child[i], word + 1)) {
            retVal = true;
            break;
        }
    }

    return retVal;
}
void wordDictionaryFree(WordDictionary* obj) {
    int i;
    for (i = 0; i < MAX_CHAR; i++) {
        if (obj->child[i]) {
            wordDictionaryFree(obj->child[i]);
        }
    }
    free(obj);
}
/**
 * Your WordDictionary struct will be instantiated and called as such:
 * WordDictionary* obj = wordDictionaryCreate();
 * wordDictionaryAddWord(obj, word);
 * bool param_2 = wordDictionarySearch(obj, word);
 * wordDictionaryFree(obj);
 */

int main(int argc, char** argv) {
#define MAX_SIZE (25 + 3)
    struct testCaseType {
        char* methods[MAX_SIZE];
        char* value[MAX_SIZE];
        int calls;
    } testCase[] = {{{"WordDictionary", "addWord", "addWord", "addWord", "search", "search", "search", "search"},
                     {"", "bad", "dad", "mad", "pad", "bad", ".ad", "b.."},
                     8}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example:
     *  Input
     *  ["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
     *  [[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
     *  Output
     *  [null,null,null,null,false,true,true,true]
     */

    WordDictionary* obj = NULL;
    bool retBool;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input\n");
        printf("[");
        for (j = 0; j < testCase[i].calls; ++j) {
            printf("%s\"%s\"", (j == 0) ? "" : ", ", testCase[i].methods[j]);
        }
        printf("]\n");
        printf("[");
        for (j = 0; j < testCase[i].calls; ++j) {
            printf("%s[\"%s\"]", (j == 0) ? "" : ", ", testCase[i].value[j]);
        }
        printf("]\n");

        printf("Output\n");
        printf("[");
        for (j = 0; j < testCase[i].calls; ++j) {
            if (strcmp(testCase[i].methods[j], "WordDictionary") == 0) {
                obj = wordDictionaryCreate();
                printf("%snull", (j == 0) ? "" : ", ");
            } else if (strcmp(testCase[i].methods[j], "addWord") == 0) {
                wordDictionaryAddWord(obj, testCase[i].value[j]);
                printf("%snull", (j == 0) ? "" : ", ");
            } else if (strcmp(testCase[i].methods[j], "search") == 0) {
                retBool = wordDictionarySearch(obj, testCase[i].value[j]);
                printf("%s%s", (j == 0) ? "" : ", ", (retBool == true ? "true" : "false"));
            }
        }
        printf("]\n");

        wordDictionaryFree(obj);
        obj = NULL;

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/211.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class TrieNode:
    def __init__(self):
        # word in addWord consists of lowercase English letters.
        self.letters = 26
        self.children = [None] * self.letters
        self.isEnd = False

    def insert(self, word: str) -> None:
        node = self
        for ch in word:
            ch = ord(ch) - ord('a')
            if not node.children[ch]:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.isEnd = True


class WordDictionary:
    def __init__(self):
        self.trieRoot = TrieNode()

    def addWord(self, word: str) -> None:
        self.trieRoot.insert(word)

    def search(self, word: str) -> bool:
        retVal = False

        wordSize = len(word)

        def dfs(index: int, node: TrieNode) -> bool:
            retVal = False

            if index == wordSize:
                retVal = node.isEnd
                return retVal

            ch = word[index]
            if ch != '.':
                child = node.children[ord(ch) - ord('a')]
                if child is not None and dfs(index + 1, child):
                    retVal = True
            else:
                for child in node.children:
                    if child is not None and dfs(index + 1, child):
                        retVal = True
                        break

            return retVal

        retVal = dfs(0, self.trieRoot)

        return retVal


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)

if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        logging.info("Input")
        for methods, val in zip([["WordDictionary", "addWord", "addWord", "addWord", "search", "search", "search", "search"]],
                                [[[], ["bad"], ["dad"], ["mad"], ["pad"], ["bad"], [".ad"], ["b.."]]]):
            # /* Example:
            #  *  Input
            #  *  ["WordDictionary", "addWord", "addWord", "addWord", "search", "search", "search", "search"]
            #  *  [[], ["bad"], ["dad"], ["mad"], ["pad"], ["bad"], [".ad"], ["b.."]]
            #  *  Output
            #  *  [null, null, null, null, false, true, true, true]
            #  */
            logging.info("%s", methods)
            logging.info("%s", val)

            displayStr = []
            for key, value in zip(methods, val):
                if key == "WordDictionary":
                    obj = WordDictionary()
                    displayStr.append(None)
                elif key == "addWord":
                    obj.addWord(value[0])
                    displayStr.append(None)
                elif key == "search":
                    retVal = obj.search(value[0])
                    displayStr.append(retVal)
            logging.info("Output")
            logging.info("%s", displayStr)

            print("")
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
